src/lm/tools/fetchTool.ts

```diff
@@ -0,0 +1,70 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import * as vscode from 'vscode';
+import { FolderRepositoryManager } from '../../github/folderRepositoryManager';
+import { IssueModel } from '../../github/issueModel';
+import { PullRequestModel } from '../../github/pullRequestModel';
+import { RepositoriesManager } from '../../github/repositoriesManager';
+
+interface FetchToolParameters {
+	issueNumber: number;
+	repo?: {
+		owner: string;
+		name: string;
+	};
+}
+
+interface FetchResult {
+	title: string;
+	body: string;
+	comments: {
+		body: string;
+	}[];
+}
+
+export class FetchTool implements vscode.LanguageModelTool<FetchToolParameters> {
+	constructor(private readonly repositoriesManager: RepositoriesManager) { }
+
+	async invoke(options: vscode.LanguageModelToolInvocationOptions<FetchToolParameters>, _token: vscode.CancellationToken): Promise<vscode.LanguageModelToolResult | undefined> {
+		const issueOrPullRequest = await fetchIssueOrPR(options, this.repositoriesManager);
+		const result: FetchResult = {
+			title: issueOrPullRequest.title,
+			body: issueOrPullRequest.body,
+			comments: issueOrPullRequest.item.comments?.map(c => ({ body: c.body })) ?? []
+		};
+		return {
+			'text/plain': JSON.stringify(result)
+		};
+	}
+}
+
+export async function fetchIssueOrPR(options: vscode.LanguageModelToolInvocationOptions<FetchToolParameters>, repositoriesManager: RepositoriesManager): Promise<PullRequestModel | IssueModel> {
```

Comment:
This seems specific to the fetch tool. Could it be a private method in the `FetchTool` class?

Reply:
Yes initially I extracted it because I wanted to reuse it for the summarization tool. My understanding however now is that we'd like to make the summarization use as input the output of the fetch tool? I am wondering if doing this could give wrong outputs sometimes as the LLM will not make the link between fetching before summarizing.

Reply:
You are correct, that can happen. In that case, we look through the previous chat messages to find relevant information. You can see an example of that here:

https://github.com/microsoft/vscode-pull-request-github/blob/abbceb87c038156493d4be3959a94e984740525a/src/lm/tools/displayIssuesTool.ts#L109-L128

To be clear, I don't know if looking through the history is best thing to do, but it's effective. I plan to see if I can tweak prompts to try to get the outputs to get passed to inputs better.